### ISSUE TITLE
Add Debian

### ISSUE THREAD

**Issue by abitrolly** [canonical/multipass#2564]
**What are you trying to do?**

I want to run Debian in a QEMU VM to create a reproducible test case.

**What's your proposed solution?**

Add Debian image to `multipass find --show-unsupported` or `multipass find` lists.

**Additional context**

Right now there is no Debian even among unsupported lists.
```
✗ multipass find --show-unsupported
Image                       Aliases           Version          Description
snapcraft:core18            18.04             20201111         Snapcraft builder for Core 18
snapcraft:core20            20.04             20210921         Snapcraft builder for Core 20
snapcraft:devel                               20220507         Snapcraft builder for the devel series
core                        core16            20200818         Ubuntu Core 16
core18                                        20211124         Ubuntu Core 18
10.04                       lucid             20150427         Ubuntu 10.04 LTS
11.10                       oneiric           20130509         Ubuntu 11.10
12.04                       precise           20170502         Ubuntu 12.04 LTS
12.10                       quantal           20140409         Ubuntu 12.10
13.04                       raring            20140111         Ubuntu 13.04
13.10                       saucy             20140709         Ubuntu 13.10
14.04                       trusty            20191107         Ubuntu 14.04 LTS
14.10                       utopic            20150723         Ubuntu 14.10
15.04                       vivid             20160203         Ubuntu 15.04
15.10                       wily              20160715         Ubuntu 15.10
16.04                       xenial            20211001         Ubuntu 16.04 LTS
16.10                       yakkety           20170719         Ubuntu 16.10
17.04                       zesty             20171208         Ubuntu 17.04
17.10                       artful            20180706         Ubuntu 17.10
18.04                       bionic            20220505         Ubuntu 18.04 LTS
18.10                       cosmic            20190628         Ubuntu 18.10
19.04                       disco             20200109         Ubuntu 19.04
19.10                       eoan              20200717         Ubuntu 19.10
20.04                       focal,lts         20220505         Ubuntu 20.04 LTS
20.10                       groovy            20210720         Ubuntu 20.10
21.04                       hirsute           20220118         Ubuntu 21.04
21.10                       impish            20220309         Ubuntu 21.10
22.04                       jammy             20220506         Ubuntu 22.04 LTS
daily:22.10                 devel,kinetic     20220510         Ubuntu 22.10
appliance:adguard-home                        20200812         Ubuntu AdGuard Home Appliance
appliance:lxd                                 20200812         Ubuntu LXD Appliance
appliance:mosquitto                           20200812         Ubuntu Mosquitto Appliance
appliance:nextcloud                           20200812         Ubuntu Nextcloud Appliance
appliance:openhab                             20200812         Ubuntu openHAB Home Appliance
appliance:plexmediaserver                     20200812         Ubuntu Plex Media Server Appliance
anbox-cloud-appliance                         latest           Anbox Cloud Appliance
charm-dev                                     latest           A development and testing environment for charmers
docker                                        latest           A Docker environment with Portainer and related tools
minikube                                      latest           minikube is local Kubernetes
```

**Comment by luis4a0** [comment]
Hi @abitrolly! Thanks for the input. The images we include in Multipass are only Ubuntu-based.

However, you can launch Debian if you are using Multipass on Linux. Just run `multipass launch file://path-to-downloaded-debian-image.iso`. Make sure that the downloaded Debian image works with cloud-init (there are many flavors of Debian images). Please also note that launching from custom images will not work on MacOS or Windows.

**Comment by abitrolly** [comment]
@luis4a0 I've used this line with the cloud image, which should be `cloud-init` ready if I understand it right. But it fails.
```shell
✗ multipass launch --name debian11 -vvvv https://cloud.debian.org/images/cloud/bullseye/latest/debian-11-generic-amd64.qcow2
[2022-05-12T20:23:41.220] [trace] [url downloader] Found https://codeload.github.com/canonical/multipass-workflows/zip/refs/heads/main in cache: true
[2022-05-12T20:23:41.246] [debug] [qemu-system-x86_64] [205317] started: qemu-system-x86_64 --version
[2022-05-12T20:23:41.941] [trace] [url downloader] Found https://cdimage.ubuntu.com/ubuntu-core/16/stable/current/SHA256SUMS in cache: true
[2022-05-12T20:23:42.147] [trace] [url downloader] Found https://cdimage.ubuntu.com/ubuntu-core/18/stable/current/SHA256SUMS in cache: true
[2022-05-12T20:23:42.376] [trace] [url downloader] Found https://cloud-images.ubuntu.com/buildd/releases/bionic/release/SHA256SUMS in cache: true
[2022-05-12T20:23:42.620] [trace] [url downloader] Found https://cloud-images.ubuntu.com/buildd/releases/focal/release/SHA256SUMS in cache: true
[2022-05-12T20:23:42.835] [trace] [url downloader] Found https://cloud-images.ubuntu.com/buildd/daily/jammy/current/SHA256SUMS in cache: true
[2022-05-12T20:23:42.837] [trace] [url downloader] Found https://cloud-images.ubuntu.com/releases/streams/v1/index.json in cache: true
[2022-05-12T20:23:42.872] [trace] [url downloader] Found https://cloud-images.ubuntu.com/releases/streams/v1/com.ubuntu.cloud:released:download.json in cache: true
[2022-05-12T20:23:44.175] [trace] [url downloader] Found https://cloud-images.ubuntu.com/daily/streams/v1/index.json in cache: false
[2022-05-12T20:23:44.177] [trace] [url downloader] Found https://cloud-images.ubuntu.com/daily/streams/v1/com.ubuntu.cloud:daily:download.json in cache: true
[2022-05-12T20:23:44.181] [trace] [url downloader] Found https://cdimage.ubuntu.com/ubuntu-core/appliances/streams/v1/index.json in cache: true
[2022-05-12T20:23:44.181] [trace] [url downloader] Found https://cdimage.ubuntu.com/ubuntu-core/appliances/streams/v1/com.ubuntu.appliances:released:download.json in cache: true
[2022-05-12T20:25:35.659] [trace] [url downloader] Found https://cloud.debian.org/images/cloud/bullseye/latest/debian-11-generic-amd64.qcow2 in cache: false
[2022-05-12T20:25:35.716] [debug] [qemu-img] [206051] started: qemu-img info --output=json /var/snap/multipass/common/cache/multipassd/vault/images/debian-11-generic-amd64-20220503/debian-11-generic-amd64.qcow2
[2022-05-12T20:25:36.528] [debug] [qemu-img] [206060] started: qemu-img info /var/snap/multipass/common/cache/multipassd/vault/images/debian-11-generic-amd64-20220503/debian-11-generic-amd64.qcow2
[2022-05-12T20:25:36.586] [debug] [qemu-img] [206065] started: qemu-img resize /var/snap/multipass/common/data/multipassd/vault/instances/debian11/debian-11-generic-amd64.qcow2 5368709120
[2022-05-12T20:25:36.835] [debug] [qemu-img] [206070] started: qemu-img snapshot -l /var/snap/multipass/common/data/multipassd/vault/instances/debian11/debian-11-generic-amd64.qcow2
[2022-05-12T20:25:36.946] [debug] [debian11] process working dir '/snap/multipass/6920/qemu'
[2022-05-12T20:25:36.947] [info] [debian11] process program 'qemu-system-x86_64'
[2022-05-12T20:25:36.947] [info] [debian11] process arguments '--enable-kvm, -cpu, host, -nic, tap,ifname=tap-ea0f558e917,script=no,downscript=no,model=virtio-net-pci,mac=52:54:00:cb:74:67, -device, virtio-scsi-pci,id=scsi0, -drive, file=/var/snap/multipass/common/data/multipassd/vault/instances/debian11/debian-11-generic-amd64.qcow2,if=none,format=qcow2,discard=unmap,id=hda, -device, scsi-hd,drive=hda,bus=scsi0.0, -smp, 1, -m, 1024M, -qmp, stdio, -chardev, null,id=char0, -serial, chardev:char0, -nographic, -cdrom, /var/snap/multipass/common/data/multipassd/vault/instances/debian11/cloud-init-config.iso'
[2022-05-12T20:25:36.975] [debug] [qemu-system-x86_64] [206092] started: qemu-system-x86_64 -nographic -dump-vmstate /tmp/multipassd.MspNRp
[2022-05-12T20:25:37.259] [info] [debian11] process state changed to Starting   
[2022-05-12T20:25:37.263] [info] [debian11] process state changed to Running    
[2022-05-12T20:25:37.264] [debug] [qemu-system-x86_64] [206109] started: qemu-system-x86_64 --enable-kvm -cpu host -nic tap,ifname=tap-ea0f558e917,script=no,downscript=no,model=virtio-net-pci,mac=52:54:00:cb:74:67 -device virtio-scsi-pci,id=scsi0 -drive file=/var/snap/multipass/common/data/multipassd/vault/instances/debian11/debian-11-generic-amd64.qcow2,if=none,format=qcow2,discard=unmap,id=hda -device scsi-hd,drive=hda,bus=scsi0.0 -smp 1 -m 1024M -qmp stdio -chardev null,id=char0 -serial chardev:char0 -nographic -cdrom /var/snap/multipass/common/data/multipassd/vault/instances/debian11/cloud-init-config.iso
[2022-05-12T20:25:37.264] [info] [debian11] process started                     
launch failed: The following errors occurred:                                   
debian11: timed out waiting for response                                                                                                                          /7m-2.2s
✗ multipass list
Name                    State             IPv4             Image
debian11                Unknown           --               Not Available                                                                                             /3.7s
✗ multipass shell 
Retrieving image: 1%^C                                                                                                                                               /8.8s
✗ multipass shell debian11
shell failed: Cannot retrieve credentials in unknown state
✗ multipass version                                                                                                         
multipass   1.9.0
multipassd  1.9.0
```

**Comment by luis4a0** [comment]
Your command line works like a charm for me. I have Debian 11 running, with Multipass 1.9.0. The issue must be somewhere else. Can you launch any Ubuntu image from the list? In any case, could you please send us the logs, as described [here](https://multipass.run/docs/accessing-logs)?

**Comment by abitrolly** [comment]
```shell
✗ sudo journalctl --unit snap.multipass*
zsh: no matches found: snap.multipass*
```
[The instruction](https://multipass.run/docs/accessing-logs) need to be fixed.
```
$ journalctl --unit "snap.multipass*" > multipass.log
```
Attached.

[debian11.log](https://github.com/canonical/multipass/files/8681498/debian11.log)

**Comment by townsend2010** [comment]
Hey @abitrolly,

Your instance is not getting an IP address.  Based on the kernel version in the logs you posted, I assume your host is running Fedora and I believe you are running into #1448.

Perhaps doing what's in https://github.com/canonical/multipass/issues/1448#issuecomment-611503315 will work, but there are other comments in there as well that have helped folks running Multipass on Fedora.

**Comment by abitrolly** [comment]
Well, yes, I am running Fedora. Sad to see that from Fedora 32 to Fedora 36 the problem is still not fixed.

What lines in log specifically tells about the the problem?

**Comment by townsend2010** [comment]
Hey @abitrolly,

I see these lines as evidence you are probably using Fedora:
```
May 12 19:59:29 blackred multipassd[193165]: Failed to enable AppArmor: AppArmor is not enabled
...
May 12 19:59:29 blackred multipassd[193165]: Cannot parse kernel version '5.17.5-300.fc36.x86_64'
```

The problem is that `ufw` is turned on by default in Fedora and will automatically block the `multipassd` daemon and the `mpqemubr0` bridge that we create for NAT'ing Multipass instances.  I'm not so sure it's a good idea for Multipass to be adding `ufw` exceptions without the user knowing.  Also, trying to account for all of the various firewalls that are out there is not a trivial task.  I think at this point, the best we could do is add a How-To to our [documentation](https://multipass.run/docs) with how to make this work for Fedora users or users who have `ufw` enabled.  I have also seen evidence that Docker and LXD are blocked by default on systems with `ufw`, so we are not alone.

**Comment by abitrolly** [comment]
@townsend2010 downloader could successfully download the image, so outgoing connection works without problems. Why accessing VM port needs NAT if I do not want to share instances with outside network? Why it is not possible to just listen on a local port?

**Comment by abitrolly** [comment]
It is still not evident that the failure is caused by firewall. If I execute the QEMU command manually, I get "Permission denied" error on downloaded `.qcow2` image.

```
✗ qemu-system-x86_64 --enable-kvm -cpu host -nic tap,ifname=tap-ea0f558e917,script=no,downscript=no,model=virtio-net-pci,mac=52:54:00:ff:9f:94 -device virtio-scsi-pci,id=scsi0 -drive file=/var/snap/multipass/common/data/multipassd/vault/instances/debian11/debian-11-generic-amd64.qcow2,if=none,format=qcow2,discard=unmap,id=hda -device scsi-hd,drive=hda,bus=scsi0.0 -smp 1 -m 1024M -qmp stdio -chardev null,id=char0 -serial chardev:char0 -nographic -cdrom /var/snap/multipass/common/data/multipassd/vault/instances/debian11/cloud-init-config.iso
qemu-system-x86_64: -drive file=/var/snap/multipass/common/data/multipassd/vault/instances/debian11/debian-11-generic-amd64.qcow2,if=none,format=qcow2,discard=unmap,id=hda: Could not open '/var/snap/multipass/common/data/multipassd/vault/instances/debian11/debian-11-generic-amd64.qcow2': Permission denied
```
The file is readable by everybody, so it is not clear where this "Permission denied" error comes from.
```
✗ ls -laZ /var/snap/multipass/common/data/multipassd/vault/instances/debian11 
total 337840
drwxr-xr-x. 2 root root system_u:object_r:snappy_var_t:s0      4096 May 31 21:19 .
drwxr-xr-x. 3 root root system_u:object_r:snappy_var_t:s0      4096 May 31 21:19 ..
-rw-r--r--. 1 root root system_u:object_r:snappy_var_t:s0     53248 May 31 21:19 cloud-init-config.iso
-rw-r--r--. 1 root root system_u:object_r:snappy_var_t:s0 346030080 May 31 21:58 debian-11-generic-amd64.qcow2
```

**Comment by townsend2010** [comment]
@abitrolly,

You need to run `qemu-system-x86_64` using `sudo` since that image is also **written** to.

**Comment by abitrolly** [comment]
@abitrolly does `multipass` really run under `root` to create those images in the first place?

**Comment by townsend2010** [comment]
>  does multipass really run under root to create those images in the first place?

Yes, `multipassd`, the daemon, runs as `root` like many other services.  One reason `root` is needed is that the `tap` device that is used for the networking has to be created by `root.  This is a requirement of Qemu, not us.

**Comment by townsend2010** [comment]
I'm going to close this as we have no plans to support Debian or other distros directly within Multipass.  We do have [plans](https://github.com/canonical/multipass/issues/307) to allow configuring custom image repositories, so I'm going to duplicate this issue to that.

Duplicate of #307

**Comment by abitrolly** [comment]
> One reason root is needed is that the tap device that is used for the networking has to be created by `root. This is a requirement of Qemu, not us.

The `tap` device is not used in SLIRP mode, right? https://wiki.qemu.org/Documentation/Networking#User_Networking_.28SLIRP.29

**Comment by wheelerlaw** [comment]
@townsend2010 

> I'm not so sure it's a good idea for Multipass to be adding `firewalld` exceptions without the user knowing.

Multipass already [does](https://github.com/canonical/multipass/blob/a1ea65fb63184fc778e277d3c9e2920ef2a9fb3d/src/platform/backends/qemu/linux/firewall_config.cpp#L151) this for `iptables` and `nftables`. The implementation for doing the same thing for `firewalld` would be much easier to write that what is already written for handling `iptables` and `nftables`.

**Comment by townsend2010** [comment]
Hey @wheelerlaw,

I've read up on `firewalld` recently and I agree with your assertion that it's not really any different than what we do already with `iptables` and `nftables`.  So what I said there isn't really true :slightly_smiling_face: 

The one thing I am concerned about is if `firewalld` would actually work in a strictly confined Snap.  `iptables` and `nftables` are in the core20 snap, so it's trivial to use those, but `firewalld` is not in the core20 snap, so we'd need to include it in our snap and then see how to allow a strictly confined `firewalld` interact with the system.

I'd love it if a developer with Fedora experience could take this on and help us out...
